### PR TITLE
 SIP2-48: Complete patron information integration

### PR DESCRIPTION
**Pull request by mreno-EBSCO**
Full end to end patron information SIP request/response.

Missing features/limitations:
- fees/fines integration
- limits
- patron status
- patron password validation
- home address is either the preferred address or the first address if none are preferred
- details returned as specified by the request's `summary` field are the item's title. This is for hold items, overdue items and recall items.
- language returned is simply what was specified by the request, but is ignored for operations

### CODE REVIEW

File: src/main/java/org/folio/edge/sip2/MainVerticle.java [hunk 1]
```diff
@@ -64,12 +66,13 @@ public void start(Future<Void> startFuture) {
           new FolioResourceProviderModule(config().getString("okapiUrl"), vertx),
           new ApplicationModule());
       handlers = new EnumMap<>(Command.class);
-      handlers.put(LOGIN, injector.getInstance(LoginHandler.class));
-      handlers.put(CHECKIN, injector.getInstance(CheckinHandler.class));
       handlers.put(CHECKOUT, injector.getInstance(CheckoutHandler.class));
```
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/MainVerticle.java]: I was ordering it to follow the SIP spec order. I think I had it that way in a few other places and wanted the order to be consistent. For some reason I thought it would make it easier to see if we were missing something. Now that I think about it, it is probably a dumb reason...

File: src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java [hunk 2]
```diff
@@ -181,12 +280,27 @@ public JsonObject getBody() {
...
+        final int offset = startItem.intValue() - 1; // expects a 1-based count, FOLIO is 0
+        sb.append("&offset=")
+          .append(offset);
+      }
+      if (endItem != null) {
+        final int limit = endItem.intValue() - (startItem != null ? startItem.intValue() - 1 : 0);
+        sb.append("&limit=")
+          .append(limit);
+      }
+
```
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java]: Hmm... But then we'd be creating a variable that may never be used. Do you think it is clearer with the variable or are you just concerned about the `intValue()` being called twice? The JIT may optimize this to inline the call. Not sure in this case.

[the remark this answers is not in the document]
Comment (mtraneis) [reply src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java]: I was more concerned with the logic of determining null being repeated. startItemValue may never be used, is that bad?

[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java]: I don't think it is that bad. I don't think sonar will complain if the variable could be used. I was just trying to isolate logic to where it is needed.

File: src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java [hunk 3]
```diff
@@ -198,12 +312,97 @@ public String getPath() {
...
+          .append("/circulation/requests?query=%28")
+          .append(idField)
+          .append("%3D%3D")
+          .append(idValue)
+          .append("%20and%20status%3DOpen");
+      if (requestType != null) {
+        sb.append("%20and%20requestType%3D%3D")
+          .append(requestType);
+      }
+      sb.append("%29");
```
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java]: Probably. The vertx web client may do it for us. I just grabbed the URLs from mod-patron and they were hand encoded, for whatever reason. I'll add it to the TODO list. 😄

File: src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java [hunk 4]
```diff
@@ -155,14 +153,115 @@
         });
   }
 
+  /**
```
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java]: I think the requirements are different for each of these workflows. I don't think the same calls are made in mod-patron. I'll have to verify. Maybe they can be. I don't think talking to edge-patron would be correct. I believe we return XML and might drop needed info. We also don't have an API key configured.

File: src/main/java/org/folio/edge/sip2/handlers/PatronInformationHandler.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        patronRepository.patronInformation(patronInformation, sessionData);
+
+    return patronFuture.compose(patronInformationResponse -> {
+      log.debug("PatronInformationResponse: {}", () -> patronInformationResponse);
+
+      final Map<String, Object> root = new HashMap<>();
+      root.put("formatDateTime", new FormatDateTimeMethodModel());
+      root.put("delimiter", sessionData.getFieldDelimiter());
+      root.put("patronInformationResponse", patronInformationResponse);
+      root.put("maxLength", 100); // should come from the sc status command
```
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/handlers/PatronInformationHandler.java]: It should probably be labelled `maxPrintWidth`. But, yes a TODO item. The intent is to limit each print line message to the max width here. Currently, we are splitting the line into smaller segments that are `maxLength` or shorter when we convert the print lines to SIP.

File: src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+    this.clock = Objects.requireNonNull(clock, "Clock cannot be null");
+  }
+
+  /**
+   * Perform patron information.
+   *
+   * @param patronInformation the patron information domain object
+   * @return the patron information response domain object
+   */
+  public Future<PatronInformationResponse> patronInformation(
```
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java]: I see your point, but I intended the names of these methods to indicate what SIP message we were executing. For example, `CirculationRepository` has `checkin` and `checkout` methods. The non SIP message methods (support) have `get` in the names. I'll have to think about this.

File: src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+            .map(o -> (JsonObject) o)
+            .findFirst()
+            .map(Optional::of)
+            .orElse(Optional.empty()))
+        .map(this::toHomeAddress);
+
+    return addressString.isPresent() ? addressString.get() : null;
+  }
+
+  private String getPatronEmailAddress(JsonObject user) {
```
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java]: Yeah, I was thinking I may need to do something specific for these fields, like personal name or home address, but I guess it turned out that I didn't. 😄 I'll refactor that.

File: src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+    final JsonObject personal = user.getJsonObject(FIELD_PERSONAL, new JsonObject());
+
+    return personal.getString("email");
+  }
+
+  private String getPatronHomePhoneNumber(JsonObject user) {
```
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java]: We could put them somewhere else, though I don't think there would be any other consumer of this logic. I'm not keen on exposing implementation to other classes. Dependencies may be formed by mistake or intentionally. I'll think about this as well.

File: src/main/resources/templates/lib.ftl [hunk 9]
```diff
@@ -83,7 +91,7 @@
 </#macro>
 
 <#macro chargedItemsCount value>
-  <@limitNumberToRange value=value min=0 max=9999 length=4/>
+  <@limitNumberToRangeOrSpaces value=value min=0 max=9999 length=4/>
```
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/resources/templates/lib.ftl]: I wouldn't trust that dev guide. According to the SIP spec, for example, hold items count:
> 4-char, fixed length field.  This field should contain the number of hold items for this patron, from 0000 to 9999. If this information is not available or unsupported this field should contain four blanks (code $20). 

The last line is why the spaces are supported. We do not support a few counts yet, so they will be spaces. If they are supported, they are 0 left padded up to the length of the field (4 in this case).

File: src/test/java/org/folio/edge/sip2/handlers/PatronInformationHandlerTests.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+    final PatronInformationHandler handler = new PatronInformationHandler(mockPatronRepository,
+        FreemarkerRepository.getInstance().getFreemarkerTemplate(PATRON_INFORMATION_RESPONSE));
+
+    final SessionData sessionData = SessionData.createSession("diku", '|', false, "IBM850");
+
+    handler.execute(patronInformation, sessionData).setHandler(
+        testContext.succeeding(sipMessage -> testContext.verify(() -> {
+          final String expectedString = "64              001|"
+              + ZonedDateTime.now(clock).plusSeconds(5)
```
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/test/java/org/folio/edge/sip2/handlers/PatronInformationHandlerTests.java]: Yup. I didn't retrofit the tests after the merge... at least not all of them. I'll go back and see if I can find all these cases.

File: src/test/java/org/folio/edge/sip2/repositories/RepositoryTestUtils.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+/**
+ * Utils for the repository tests.
+ *
+ * @author mreno-EBSCO
+ *
+ */
```
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/test/java/org/folio/edge/sip2/repositories/RepositoryTestUtils.java]: Yeah, I wrote this before I merged in your PR. In fact, I see that I forgot to change the class name from `PatronRepositoryTests` to `RepositoryTestUtils` when I moved it out of the class. 😞

File: src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+            holds -> addHolds(holds, patronInformation.getSummary() == HOLD_ITEMS, builder));
+    final Future<PatronInformationResponseBuilder> overdueFuture =
+        circulationRepository.getOverdueLoansByUserId(userId, ZonedDateTime.now(clock),
+            startItem, endItem, sessionData).map(
+                overdues -> addOverdueItems(overdues,
+                    patronInformation.getSummary() == OVERDUE_ITEMS, builder));
+    final Future<PatronInformationResponseBuilder> recallsFuture =
+        getRecalls(userId, sessionData).compose(recalls -> addRecalls(recalls, startItem, endItem,
+            patronInformation.getSummary() == RECALL_ITEMS, builder));
+    return CompositeFuture.all(holdsFuture, overdueFuture, recallsFuture)
```
Comment (mtraneis) [inline src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java]: More lambda magic... How do the results of holds/overdue/recallsFutures getting used by the builder?

Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java]: In this case, there is not too much magic. I'll add some comments to describe the workflow here, but basically, each async call has the same builder and adds the items it is responsible for to the builder, holds (count and items, if requested), overdues (same as holds) and recalls (same as holds, but assembled differently). When all the futures complete (`CompositeFuture.all()`), we add the final items and build the response.

Because of the complexity (over 60) of the `patronInformation` method I had to split it up quite a bit, which actually simplified some of the logic that was in there. It is still a bit complicated.

File: src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java]: Maybe we could add a JSON utility class to help with common tasks. I am noticing I am doing a lot of the same things, like:
`jsonObject.getJsonObject("some-object", new JsonObject()).getString("some-field")`
Which basically means that I want to get a field within a child object, but if the object is not present, I don't want a NPE, so build a new `JsonObject` before calling `getString`. Since the object should be eagerly created as a method argument, it would probably be better to check for `null` first. That would make the code look like blah, so I'll just move these types of operations to a JSON utility class.

File: src/main/resources/templates/lib.ftl (hunk 9 above)
[the remark this answers is not in the document]
Comment (mtraneis) [reply src/main/resources/templates/lib.ftl]: I agree with you that the dev guide can't be trusted completely. However, I interpret this statement differently. It goes to length to show the range of 0000 to 9999, which signals to me that it wants the numbers 0's padded. Also, the second sentence tells me that if (and only if) this information is not available then this field should contain four blanks.  Not sure if we have another source to get guidance from. *sigh* What a spec!

File: src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mtraneis) [reply src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java]: Yeah, this plays trick on my mind, or the reader's mind. _checkin_/_checkout_ appears to be verbs, while patronInformation is a noun, plus we have other patronInformation such as a type or variable name in other places, so it was hard to follow.

File: src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mtraneis) [reply src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java]: That's why we have a story for edge-patron to also return json when asked for.  A REST framework should be able to take care of this switching for us for free, at least .NET does. I know it's a stretch, but it's something to think about for the sake of reuse.  Perhaps in the future we could add more content to the edge-patron getPatronAccount response and allows users to choose which pieces of information they want in the response.   Just something to ponder.

File: src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mtraneis) [reply src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java]: so if the web client may do it for us, then it would double encode the query string, no?  Have you tested end-to-end to see if it would work? 

If there's only one PR comment that you could address due to time constraint, I'd love to see this one being addressed because it's a sore sight to see hard-coded URL encoded strings. Not only that, it would be difficult to maintain/debug. But if there is bigger fish to fry at this moment, this can wait. My two cents  :)

File: src/main/java/org/folio/edge/sip2/MainVerticle.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mtraneis) [reply src/main/java/org/folio/edge/sip2/MainVerticle.java]: No worries. Just want to make sure that I'm not missing out on any important implicit concepts.

File: src/main/resources/templates/lib.ftl (hunk 9 above)
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/resources/templates/lib.ftl]: The macro you cite **does** pad the number with '0's if it is less than the max length. The only thing I added was to use spaces when the value is null (unsupported by ACS) to make it compliant with the spec, which is why I changed the macro from the original which pads to the fixed length with '0' to the new one that also pads with '0' if not null or uses all spaces when null. I believe only the "count" fields in the Patron Information Response have this twist.

```
<#macro limitNumberToRangeOrSpaces value min max length>
  <#if value?has_content>
    <@limitNumberToRange value=value min=min max=max length=length/><#t>
  <#else><#rt>
    ${value?right_pad(length)}<#lt><#rt>
  </#if>
</#macro>

<#macro limitNumberToRange value min max length>
  <#if value gt max>
    ${max?c}<#t>
  <#elseif value lt min>
    ${min?c?left_pad(length, "0")}<#t>
  <#else>
    ${value?c?left_pad(length, "0")}<#t>
  </#if>
</#macro>
```

File: src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java]: The problem is that there is no guarantee edge/mod-patron is going to be deployed. I think we should keep the dependencies on the core modules for now. A self hosting library may not want to deal with managing API keys or installing more modules than necessary. It is possible we could move this logic to mod-patron though. I think if that module is more of a generic patron services module, than EDS integration, then we could probably consolidate these calls. Though, I am not sure that the APIs will line up for SIP2 and EDS, but I agree we could (should) give it a try at some point.

File: src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java]: Yeah, the naming wasn't ideal. The intent was `patronInformation` was an action to perform that was defined in the SIP spec, so I wanted the names to match for ease of use, but it turned out to be the opposite. I'll rename these "action" methods to something like `perform<SIP-Command>Command`, so `performPatronInformationCommand` or `performCheckinCommand`.

File: src/main/resources/templates/lib.ftl (hunk 9 above)
[the remark this answers is not in the document]
Comment (mtraneis) [reply src/main/resources/templates/lib.ftl]: Ah, I misunderstood your previous response. Cool! Didn't realize we have been on the same page.

File: src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java]: It may be a flag or it might autodetect non-URL chars an encode them, leaving the existing encoding in place or it might do none of these. I have tested end to end and that works. I'll take a look into this next.

[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java]: It turns out vertx uses netty to manage the url encoding of parameters, so we don't need to pre-encode, like the code mentioned here. The netty encoder takes care of the '+' encoding for spaces that the Java url encoder uses by replacing them with %20. I've removed the encoding in the strings and things are working as expected.
